packages/expo-notifications/ios/EXNotifications/EXInstallationIdProvider.m

```diff
@@ -15,12 +16,97 @@ @implementation EXInstallationIdProvider
 
 - (NSString *)getInstallationId
 {
-  NSString *uuid = [[NSUserDefaults standardUserDefaults] stringForKey:kEXDeviceInstallUUIDKey];
-  if (!uuid) {
-    uuid = [[NSUUID UUID] UUIDString];
-    [[NSUserDefaults standardUserDefaults] setObject:uuid forKey:kEXDeviceInstallUUIDKey];
+  NSString *installationId = [self fetchInstallationId];
+  if (installationId) {
+    return installationId;
   }
-  return uuid;
+  
+  installationId = [[NSUUID UUID] UUIDString];
+  [self setInstallationId:installationId];
+  return installationId;
+}
+
+- (nullable NSString *)fetchInstallationId
+{
+  NSString *installationId;
+  CFTypeRef keychainResult = NULL;
+  
+  if (SecItemCopyMatching((__bridge CFDictionaryRef)[self installationIdGetQuery], &keychainResult) == noErr) {
+    NSData *result = (__bridge_transfer NSData *)keychainResult;
+    NSString *value = [[NSString alloc] initWithData:result
+                                            encoding:NSUTF8StringEncoding];
+    // `initWithUUIDString` returns nil if string is not a valid UUID
+    if ([[NSUUID alloc] initWithUUIDString:value]) {
+      installationId = value;
+    }
+  }
+  
+  if (installationId) {
+    return installationId;
+  }
+  
+  NSString *legacyUUID = [[NSUserDefaults standardUserDefaults] stringForKey:kEXDeviceInstallationUUIDLegacyKey];
+  if (legacyUUID) {
+    installationId = legacyUUID;
+
+    NSError *error = nil;
```

Comment:
BTW this is fine (and is how I've written this kind of code) but it's also correct to write `NSError *error;`. Unlike C, ARC will initialize object pointers to `nil` even on the stack.

Reply:
Initially I've written it like `NSError *error;` knowing it will get initialized to `nil`, but then I decided to see how this structure has been used in the codebase in other places to follow existing conventions and I saw https://github.com/expo/expo/blob/57a8ea27d2e83dc4d749c79fdf5419849e33db7f/ios/Exponent/Kernel/Api/EXApiV2Client.m#L98 and decided to follow its example. Thanks for the note and confirming my expectations!